specification.md

```diff
@@ -1063,6 +1067,19 @@ resources reserved through the advanced reservation represented by this
 ID.
 
 
+<a name="jobattributes-setcustomattribute"></a>
+```java
+void setCustomAttribute(String name, Object value);
+Object? getCustomAttribute(String name);
+```
+
+Allows setting/querying of custom attributes. Implementations should use a
+dictionary to implement the custom attributes. Additionally, implementations are
+encouraged to make sensible decisions on whether to store some or all of the
+fixed attributes in the same dictionary or not. It is, therefore, entirely
+possible for `getCustomAttribute("duration")` to return a value passed earlier
+to `setDuration()`, although the specific custom attribute name need not be
+`"duration"`.
 
```

Comment:
I would suggest not to speciofy the data storage type (dictionary) - the API allows any internal data storage, and I don't see a benefit in limiting that freedom.

As for the overlap with defined attributes, I see your point, but wonder if it wouldn't be easier for the user if no such overlap exists?  Because it implies also the inverted case: that the setting of a special attribute changes one of the other specified attribute's values (such as duration).  That sounds like surprising behaviour, doesn't it?  I also would worry to further impact portability, as it basically means that applications could code directly against special attributes w/o caring about the standardized ones - then why do we specify `duration` in the first place?

Reply:
My worry is that, without specifying that, the whole thing could be interpreted to require storage of the two in different places, which may affect the freedom of implementations to make certain optimizations. Basically, this is saying that we don't mandate it one way or the other and implementations are free to pick whichever solution they want.

But I see your point when it comes to the "custom" label on some of the attributes. That word isn't strictly necessary in the API and setAttribute(name, value)/getAttribute(name) are perfectly fine with me. Do you think removing "custom" would be a better choice?

Reply:
Is `Implementations are encouraged to make sensible decisions...` a useful statement?  I mean, we can make that statement on every single line of  code ;-)

As for the labels, I would suggest the following:
```
An implementation is not allowed to support custom attributes which are semantically equivalent to the attributes statically defined above (`Duration`, `QueueName`, `ProjectName`, `ReservationID`).  This removes potential confusion, and forces use of the proper API attributes.

Reply:
> This removes potential confusion, and forces use of the proper API attributes.

Except that isn't what I want. I want the implementations to have the flexibility of doing that, too. For example, I want it to be possible to implement ALL attributes as a single dictionary and, for languages that support it, also have type-safe methods for putting/getting specific known attributes.

Reply:
> Except that isn't what I want. I want the implementations to have the flexibility of doing that, too. For example, I want it to be possible to implement ALL attributes as a single dictionary and, for languages that support it, also have type-safe methods for putting/getting specific known attributes.

That's a language binding issue.  In Python for example, both defined and custom attributes can be exposed as class properties, either one can have type checks or not, and the constructor can still have a dictionary initializer.  For languages which don't have that, I am not sure that having different, potentially conflicting routes to set a well defined attribute would be good.

I won't block the PR about that issue, I think you heard my point by now.

Reply:
> I am not sure that having different, potentially conflicting routes to set a well defined attribute would be good.

It is somewhat common, in statically typed languages, to have properties with potentially unknown names be implemented using dictionaries while also providing typed accessor methods for the subset of properties that is known. I want to make it clear that the spec does not prevent such an implementation from happening.

I think I still want to consider renaming setCustomAttribute/getCustomAttribute to setAttribute/getAttribute. What do you think?

Reply:
> Because it implies also the inverted case: that the setting of a special attribute changes one of the other specified attribute's values (such as duration). That sounds like surprising behaviour, doesn't it? I also would worry to further impact portability, as it basically means that applications could code directly against special attributes w/o caring about the standardized ones - then why do we specify duration in the first place?

> For languages which don't have that, I am not sure that having different, potentially conflicting routes to set a well defined attribute would be good.

Just wanted to chime in to say that I agree.  I like the idea of the custom attributes but not as an alternate way to set/get values that we already have interfaces for.  Seems like this is asking for incompatibility across implementations and surprising behavior to users.